### DISCUSSION TITLE
How to specify a .NET 4.8 runtime in pythonnet 3.x?

### DISCUSSION THREAD

**Discussion by pd185112** [pythonnet/pythonnet#2005] [Q&A]
I'm trying to use pythonnet.3.0.1 to load/use a DLL in Python code that is compiled to require a .NET 4.8 (Target Framework: 4.8). Loading DLL doesn't work (ModuleNotFoundError: No module named 'DllThatRequires4.8'). I'm able to load a DDL from the same path that doesn't require .NET 4.8 (Target Framework: 4.0). 

I suspect my issue is that pythonnet doesn't target the 4.8 runtime (see undefined runtime version in code output), how can I configure it to do so? I've found some examples of loading a  .NET core with config: https://learn.microsoft.com/en-us/dotnet/core/runtime-config/#runtimeconfigjson, but nothing on legacy .NET.

My Windows Server 2016 machine has a .NET 4.8 installed, as shown under HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\NET Framework Setup\NDP\v4\Full\Version: 4.8.03761
Using Windows server 2016, python 3.10.8 32-bit, pythonnet 3.0.1

My code (simplified):
from pythonnet import get_runtime_info
sys.path.append('path\\to\\dll\folder')
import clr
print(get_runtime_info())
clr.AddReference('DllThatRequires4.8')
from DllThatRequires4.8 import something #this doesn't work
clr.AddReference('DllThatDoesNotRequire4.8')
from DllThatDoesNotRequire4.8 import something # this works

code output:
![image](https://user-images.githubusercontent.com/68049578/200595053-1018f45d-d745-4cf0-bb1c-d65dbf162f88.png)

....
File "steps\Automation\client.py", line 8, in <module>
    from DllThatRequires4.8 import something 
ModuleNotFoundError: No module named 'DllThatRequires4.8'

**Comment by filmor** [comment]
The runtime version is just not implemented (yet) for .NET Framework. If .NET 4.8 is installed, it should be used (I think :)). You could try to modify `clr_loader` here, https://github.com/pythonnet/clr-loader/blob/master/netfx_loader/ClrLoader.csproj, to require `net48`.

**Reply by pd185112** [reply]
Curious thing though, the setup described above is an attempt to upgrade the test environment running on Python 3.7.9/pythonnet 2.4.0 for a good while. This outdated setup works with 4.8 DLLs as expected, and I tried to jump versions directly to Python 3.10.8 32-bit/ pythonnet 3.0.1. 
When I tried Python 3.7.9/pythonnet 2.5.0 combo to run the code passing on 2.4.0, the issue described above was reproduced. So it's not a problem exclusive to 3.0.x as I originally assumed.

**Reply by filmor** [reply]
Hmm, I don't think there were any significant changes between 2.4 and 2.5, just cleanups and additional code, see https://github.com/pythonnet/pythonnet/compare/v2.4.0...v2.5.0. In particular, none of the changes touch the target frameworks, as far as I can tell. Can you try to create a minimal DLL that exhibits the behaviour?

**Reply by pd185112** [reply]
Sorry for delay in replying. I, along with some colleagues, were trying to find the root cause, but no joy so far. 
You're correct, minimum DLLs didn't confirm the original assumption with .NET version. Besides, I found one other production library that requires .NET 4.8 and that one is loaded by pythonnet 3.x as expected. 
The issue still persists with production libraries though. I had to switch to another task for the moment, but we'll return to this in the next sprint. If you happen to have any other theories, feel free to suggest things to try out...

**Reply by YoshiyasuIzumi** [reply]
Sorry for my interruption, but I assume I also faced same issue. In my case I prepared wrapper C# DLL to be called from Pythonnet as workaround. I understand this is not best way but I hope this help.
https://github.com/pythonnet/pythonnet/discussions/2558#discussioncomment-12874234
